runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaPublishFactory.java

```diff
@@ -303,12 +305,39 @@ private void onMqttBegin(
                 .value(topicNameBuffer, 0, topicNameBuffer.capacity())
                 .build();
 
-            messages.doKafkaBegin(traceId, authorization, affinity, kafkaMessagesTopic);
+            final String16FW clientHashKey = clientHashKey(topicName);
+            final DirectBuffer clientHashKeyBuffer = clientHashKey.value();
+            if (clientHashKeyBuffer != null)
+            {
+                final MutableDirectBuffer hashKeyBuffer = new UnsafeBuffer(new byte[clientHashKeyBuffer.capacity() + 4]);
+                hashKey = new KafkaKeyFW.Builder()
+                    .wrap(hashKeyBuffer, 0, hashKeyBuffer.capacity())
+                    .length(clientHashKeyBuffer.capacity())
+                    .value(clientHashKeyBuffer, 0, clientHashKeyBuffer.capacity())
+                    .build();
+            }
+
+            messages.doKafkaBegin(traceId, authorization, affinity);
             this.retainAvailable = (mqttPublishBeginEx.flags() & 1 << MqttPublishFlags.RETAIN.value()) != 0;
             if (retainAvailable)
             {
-                retained.doKafkaBegin(traceId, authorization, affinity, kafkaRetainedTopic);
+                retained.doKafkaBegin(traceId, authorization, affinity);
+            }
+        }
+
+        private String16FW clientHashKey(
+            String topicName)
+        {
+            String clientHashKey = null;
+            if (clients != null)
+            {
+                for (Function<String, String> client : clients)
+                {
+                    clientHashKey = client.apply(topicName);
+                    break;
+                }
             }
+            return new String16FW(clientHashKey);
```

Comment:
Return `null` when `clients` is `null` to avoid unused `String16FW` allocation. Call site then needs to check for `null` directly instead of for `.value()` buffer.